File: Python/150. Evaluate Reverse Polish Notation.py
import collections
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        q = collections.deque([])
        operators = set(['+', '-', '*', '/'])

        for op in tokens:
            if op not in operators:
                q.append(int(op))
            else:
                n2 = q.pop()
                n1 = q.pop()
                if op == '+':
                    q.append(n1+n2)
                elif op == '-':
                    q.append(n1-n2)
                elif op == '*':
                    q.append(n1*n2)
                elif op == '/':
                    n = int(n1/n2)
                    if n < 0 and n1 % n2: n += 1
                    q.append(n)

        return q.pop()

File: CPP/150. Evaluate Reverse Polish Notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        vector<string> operators = {"+", "-", "*", "/"};
        stack<int> stack;
        int n;
        for (string& s: tokens){
            if (find(operators.begin(), operators.end(), s) == operators.end()){
                n = stoi(s);
                stack.push(n);
            }
            else{
                int n2 = stack.top();
                stack.pop();
                int n1 = stack.top();
                stack.pop();
                if (s == "+") n = n1+n2;
                else if (s == "-") n = n1-n2;
                else if (s == "*") n = n1*n2;
                else if (s == "/") n = n1/n2;
                stack.push(n);
            }
        }
        return stack.top();
    }
};